infra/lambda_functions/websocket/disconnect/app.js

```diff
@@ -44,43 +49,113 @@ exports.handler = async (event) => {
-        );
-        await ddb.delete(deleteParamsOpenDiagramConnection).promise();
-        console.log("Successfully deleted connection:", connectionToRemove);
-      } catch (e) {
-        console.log("Error when deleting connection: ", e);
-      }
-    }
+  const connectionToRemove = connectionsToRemove.Items[0];
+  if (!connectionToRemove) {
+    console.log("No connection to remove from open diagrams table");
+    // return {
+    //   statusCode: 200,
+    //   body: "Disconnected",
+    // };
+  }
+  console.log("Attempting to remove connection:", connectionToRemove);
+
+  const deleteParamsOpenDiagramConnection = {
+    TableName: OPEN_DIAGRAMS_TABLE_NAME,
+    Key: {
+      connectionId: connectionToRemove.connectionId,
+      diagramId: connectionToRemove.diagramId,
+    },
+  };
+  try {
+    console.log("deleteParamsConnection:", deleteParamsOpenDiagramConnection);
+    await ddb.delete(deleteParamsOpenDiagramConnection).promise();
+    console.log("Successfully deleted connection:", connectionToRemove);
   } catch (e) {
     console.log("Error when deleting connection: ", e);
-    return {
-      statusCode: 500,
-      body: "Failed to disconnect: " + JSON.stringify(e),
-    };
   }
 
-  return { statusCode: 200, body: "Disconnected." };
+  console.log(
+    "We are about to check if we need a new master for:",
+    connectionToRemove
+  );
+
+  if (connectionToRemove.isMaster) {
+    console.log("we need to choose a new master");
+    await chooseMaster(ddb, connectionToRemove.diagramId, apigwManagementApi);
+  } else {
+    console.log("We do not a new master");
+  }
+
+  // return { statusCode: 200, body: "Disconnected." };
 };
+
+async function chooseMaster(ddb, diagramId, apigwManagementApi) {
+  /* 
+  TODO: if the newly-chosen master can't be reached, 
+   it needs to be deleted and this function should retry
+  */
+  let usersOnDiagramResult;
+  try {
+    usersOnDiagramResult = await ddb
+      .query({
+        TableName: OPEN_DIAGRAMS_TABLE_NAME,
+        KeyConditionExpression: "diagramId = :d",
+        ExpressionAttributeValues: {
+          ":d": diagramId,
+        },
+      })
+      .promise();
+  } catch (e) {
+    console.log("Error when querying the users on the open diagram: ", e);
+    // return { statusCode: 500, body: e.stack };
+  }
+
+  if (usersOnDiagramResult.Items.length === 0) {
+    return;
+  }
+
+  const newMasterIndex = Math.floor(
+    Math.random() * usersOnDiagramResult.Items.length
+  );
+  const newMasterUserData = usersOnDiagramResult.Items[newMasterIndex];
+  try {
+    await apigwManagementApi
+      .postToConnection({
+        ConnectionId: newMasterUserData.connectionId,
+        Data: JSON.stringify({
+          type: "master",
+        }),
+      })
+      .promise();
+  } catch (e) {
+    console.log("Failed to notify user they are master");
+    // return { statusCode: 500, body: "User dropped off" };
+  }
+
+  try {
+    await ddb
+      .put({
```

Comment:
We update the db after notifying the user... This is a bit odd... The db is supposed to be the source of truth, right? So what if we've told the user they're master but the write to the db failed. We could then end up with multiple clients thinking that they're masters.... This is actually a good use-case for dynamodb streams. First we update the db to say who is the new master, that then triggers a lambda function, which is responsible for notifying the user. If the user can't be reached, then we call disconnect on them and this process restarts. What do you think?